### ISSUE TITLE
[data] records

### ISSUE THREAD

**Issue by fwbrasil** [getkyo/kyo#917]
I've explored an encoding of records using a type union to maintain the structure information. It's able to provide proper sub-type relationships between records but I couldn't figure out an issue. If a field is added multiple times with different types, they should be unified with a type union of the possible value types. Adding variance to `Value` doesn't work because then getting values from the record get inferred as `Nothing`.

Another option is macros but I wanted to see if it'd be possible to avoid them.

https://scastie.scala-lang.org/46EN1AHCQwOEpvI6YTXo9Q
```scala
object impl:

    infix opaque type ~>[Name <: String, Value] = (Name, Value)

    extension [Name <: String](self: Name)
        def ~>[Value](value: Value): self.type ~> Value = (self, value)

    opaque type Record[-Fields] = Map[String, Any]

    object Record:
        def init: Record[Nothing] = Map.empty

        extension [Fields](self: Record[Fields])
            def toMap: Map[String, Any] = self

            def apply[Name <: String, Value](field: Name ~> Value): Record[Fields | Name ~> Value] =
                self + field

            def apply[Name <: String, Value](name: Name)(using name.type ~> Value <:< Fields): Value =
                self(name).asInstanceOf[Value]

            def ++[Fields2](other: Record[Fields2]): Record[Fields | Fields2] =
                self ++ other
        end extension

    end Record
end impl

object Demo:
    import impl.*

    type Person   = Record["name" ~> String | "age" ~> Int]
    type Employee = Record["name" ~> String | "age" ~> Int | "department" ~> String]
    type Client   = Record["name" ~> String | "age" ~> Int | "company" ~> String]

    val employee: Employee       = Record.init("name" ~> "John")("age" ~> 33)("department" ~> "some dept")
    val employeeIsPerson: Person = employee

    val person: Person = Record.init("name" ~> "John")("age" ~> 33)
    // val personIsNotEmployee: Employee = person

    // val cantAddFieds: Record["name" ~> String | "inexistent" ~> Int] = person

    // This should be Record["field" ~> (String | Int)]
    val bug: Record["field" ~> String & "field" ~> Int] = Record.init("field" ~> "someString")("field" ~> 42)
end Demo
```

cc/ @johnhungerford in case you want to have some fun with a type-level puzzle :)

**Comment by HollandDM** [comment]
purely type-level approach, AFAIK, will need `Fields <: Tuple`, then we can do logic on the Tuple itself

**Comment by johnhungerford** [comment]
I didn't have any success using a compiletime inlining approach. I may try again later using implicit resolution. My experience, however, is that you can't match types in intersections and unions reliably. I suspect the only way to make that work properly is as @HollandDM suggests, to replace with a tuple, which I don't recommend.

I was able to make it *almost* work by making `Value` covariant and then setting some restrictions using `NotGiven`.  It works when you explicitly annotate your types, but for some reason it does not infer correctly. It won't automatically turn `Record["field" ~> T1 | "field" ~> T2]` into `Record["field" ~> (T1 | T2)]`. See here: https://scastie.scala-lang.org/MfsohYNbQQGiRWWQYNKiFA

**Comment by johnhungerford** [comment]
Yeah can't make it work with implicit resolution either. Macro might be needed here.

**Comment by fwbrasil** [comment]
Thanks for giving it a try :) I've realized something: we could use `Tag` to allow repeated fields with different types:

https://scastie.scala-lang.org/Ty3ST7GXQI2nPLSa0dking
```scala
import Record.Field

infix opaque type ~>[Name <: String, Value] = (Field[Name, Value], Value)

extension [Name <: String](self: Name)
    def ~>[Value](value: Value)(using tag: Tag[Value]): self.type ~> Value = (Field(self, tag), value)

opaque type Record[-Fields] = Map[Field[?, ?], Any]

object Record:
    case class Field[Name <: String, Value](name: Name, tag: Tag[Value])

    def init: Record[Nothing] = Map.empty

    extension [Fields](self: Record[Fields])
        def apply[Name <: String, Value](entry: Name ~> Value): Record[Fields | Name ~> Value] =
            self + entry

        def apply[Name <: String](name: Name)[Value](using ev: name.type ~> Value <:< Fields, tag: Tag[Value]): Value =
            self(Field(name, tag)).asInstanceOf[Value]

        def ++[Fields2](other: Record[Fields2]): Record[Fields | Fields2] =
            self ++ other

        def toMap: Map[Field[?, ?], Any] = self
    end extension

end Record
```

@johnhungerford Do you know if it's possible to use mirrors to provide conversions between records and case classes? I took a look at the API but it seems it doesn't offer a way to produce the final record type based on the case class. I can't find a way to obtain the field names for example other than at runtime.

**Comment by johnhungerford** [comment]
Yes. See POC here:
https://scastie.scala-lang.org/YhdQY5gaT9mJ0dlz8rDTAg

**Comment by johnhungerford** [comment]
Better version: https://scastie.scala-lang.org/uH7BvmmTQ2CnvAeji4g94A (the one above does not construct the underlying `Map[Field[?,?], Any]` correctly, though it gets the types right).

@fwbrasil Do you have any particular plans for using this in Kyo?

**Comment by fwbrasil** [comment]
cool, thanks for looking into it!! Yes, I'd like to add it in `kyo-data`. I wanted to use it for a project I'm working on. I keep having to create case classes to merge information from other case classes. Records would be more convenient. Do you think it'd be a good addition?

**Comment by johnhungerford** [comment]
It's definitely a cool utility. For my own work I've been trying to work out a way to define a "fields-first" data model, and this seems like a promising approach. (I.e., define fields your fields, and then compose them to define records. What I really want is for the primary component of the data model to be an accessor like in zio-schema.)

I just don't see where it fits into Kyo as a whole. I thought maybe you had some kind of effect in mind.

**Comment by johnhungerford** [comment]
Have you looked at this?: https://docs.scala-lang.org/scala3/book/types-structural.html

Seems like something similar could be accomplished using structural types and `selectDynamic`

**Comment by fwbrasil** [comment]
> I just don't see where it fits into Kyo as a whole. I thought maybe you had some kind of effect in mind.

The idea came from an [experiment](https://github.com/getkyo/kyo/blob/hdl-experiment/kyo-direct/shared/src/main/scala/kyo/Chip.scala) to use a Kyo effect as a more flexible "front end" for [Chisel](https://www.chisel-lang.org/) that integrates via FIRRTL. I don't have plans to work on it, though. I guess we could use it in `Var` to track named vars instead of only via the type but I'm not sure it'd be a good improvement given the syntax overhead.

> Have you looked at this?: https://docs.scala-lang.org/scala3/book/types-structural.html
> 
> Seems like something similar could be accomplished using structural types and `selectDynamic`

Yes, but it doesn't provide proper inference and requires manually defining the structural type and casting records to it. I imagine we could hide that via macros but I think a solution leveraging the type system itself via union types would provide a better experience.

**Comment by sideeffffect** [comment]
>> structural types and `selectDynamic`
>
>Yes, but it doesn't provide proper inference and requires manually defining the structural type and casting records to it

Could the recently introduced _named tuples_ help us better than _structural types_? https://docs3.scala-lang.org/sips/named-tuples.html

**Comment by fwbrasil** [comment]
> > > structural types and `selectDynamic`
> > 
> > 
> > Yes, but it doesn't provide proper inference and requires manually defining the structural type and casting records to it
> 
> Could the recently introduced _named tuples_ help us better than _structural types_? https://docs3.scala-lang.org/sips/named-tuples.html

It might be a replacement for `~>` but the goal of `Record` is to provide sub-type relationship between instances so you can pass a record with more fields to a method that uses only subset of them. Named tuples don't provide that and are even sensitive to field ordering: https://docs3.scala-lang.org/sips/named-tuples.html#conformance-and-convertibility

**Comment by HollandDM** [comment]
Does record4s work for your cases

**Comment by fwbrasil** [comment]
> Does record4s work for your case?

I don't know how I missed this library! It seems great, I'll try using it for now. I think we can still keep this open since a type-level approach might provide better inference and we could reuse it more easily to, for example, implement a named `Var` effect. I'm not sure about this specific use case but I imagine we'll identify more compelling applications.
